lib/foreman_inventory_upload/generators/fact_helpers.rb

```diff
@@ -57,55 +57,187 @@ def cloud_provider(host)
+
+        # Retrieve the host-specific fact related to hostname obfuscation from 'insights_client'.
+        # This fact indicates whether hostname obfuscation is enabled for this particular host.
         insights_client_setting = fact_value(host, 'insights_client::obfuscate_hostname_enabled')
+
+        # Cast the retrieved setting to a boolean value.
+        # This ensures that the value from the fact (which might be a string or other type)
+        # is correctly interpreted as true or false.
         insights_client_setting = ActiveModel::Type::Boolean.new.cast(insights_client_setting)
+
+        # If the host-specific setting is not nil (meaning it was explicitly set to true or false),
+        # return its boolean value directly. This prioritizes the host-specific setting
+        # over the global default if the host has a specific configuration.
         return insights_client_setting unless insights_client_setting.nil?
 
+        # If the host-specific setting was nil (meaning it wasn't explicitly set for this host),
+        # fall back to the global 'obfuscate_inventory_hostnames' setting.
+        # This serves as the default behavior when no specific host configuration is found.
         Setting[:obfuscate_inventory_hostnames]
       end
 
       def fqdn(host)
-        return host.fqdn unless obfuscate_hostname?(host)
-
-        fact_value(host, 'insights_client::hostname') || obfuscate_fqdn(host.fqdn)
+        # Returns the Fully Qualified Domain Name (FQDN) for a given host,
+        # potentially obfuscating it if hostname obfuscation is enabled.
+        # Check if hostname obfuscation is enabled for this host using the `obfuscate_hostname?` method.
+        if obfuscate_hostname?(host)
+          # If obfuscation is enabled, attempt to retrieve an already obfuscated hostname
+          # from the 'insights_client::obfuscated_hostname' fact.
+          # This fact is expected to contain a JSON array of mappings between original and obfuscated hostnames.
+
+          # Parse the 'insights_client::obfuscated_hostname' fact value as a JSON array.
+          # If the fact is not present or empty, default to an empty array to prevent parsing errors.
+          # Example format of `parsed_insights_array`:
+          # [{"original"=>"host.example.com", "obfuscated"=>"0dd449d0a027.example.com"},
+          #  {"original"=>"satellite.example.com", "obfuscated"=>"host2.example.com"}]
+          begin
+            parsed_insights_array = JSON.parse(fact_value(host, 'insights_client::obfuscated_hostname') || '[]')
+          rescue JSON::ParserError
+            # If the fact contains invalid JSON, default to an empty array to prevent upload failures.
+            parsed_insights_array = []
+          end
+
+          # Find the specific item in the parsed array where the 'original' hostname matches
+          # the host's actual FQDN.
+          parsed_insights_item = parsed_insights_array.find { |item| item['original'] == host.fqdn }
```

Comment:
I think the `find` is pretty self-explanatory here
```suggestion
          parsed_insights_item = parsed_insights_array.find { |item| item['original'] == host.fqdn }
```